File: 백준/Gold/12944. 재미있는 숫자 놀이/재미있는 숫자 놀이.py
def gcd(x, y):
    if x < y:
        x, y = y, x
    while y > 0:
        x, y = y, x % y
    return x


def combination(current, index, total):
    global answer
    lcm = 1

    for i in range(current):
        lcm /= gcd(lcm, select[i])
        lcm *= select[i]

    if current > 0:
        answer = answer + int(N/lcm) if (current & 1) == 1 else answer - int(N/lcm)

    if current == total:
        return

    for i in range(index, K):
        select[current] = A[i]
        combination(current+1, i+1, total)


select = [0] * 20
answer = 0
N, K = map(int, input().split())
A = list(map(int, input().split()))
combination(0, 0, K)
print(int(answer))

File: 백준/Gold/12944. 재미있는 숫자 놀이/재미있는 숫자 놀이.c
#include<stdio.h>

long long N, A[20], select[20], answer=0;
int K;

long long gcd(unsigned long long x,unsigned long long y)
{
	long long temp;

	if(x<y)
	{
		temp=x;
		x=y;
		y=temp;
	}

	while(y)
	{
		temp=x%y;
		x=y;
		y=temp;
	}

	return x;
}

void combination(int current,int index,int total)
{
	long long lcm=1;

	for(int i=0;i<current;++i)
	{
		lcm/=gcd(lcm,select[i]);
		lcm*=select[i];

		if(lcm>1000000000)
			break;
	}

	if(current && lcm<=1000000000)
	{
		if(current&1)
			answer+=N/lcm;
		else
			answer-=N/lcm;
	}

	if(current==total || lcm>1000000000)
		return;

	for(int i=index;i<K;++i)
	{
		select[current]=A[i];
		combination(current+1,i+1,total);
	}
}

int main(void)
{
	scanf("%lld%d", &N, &K);

	for(int i=0;i<K;++i)
		scanf("%lld", &A[i]);

	combination(0,0,K);

	printf("%lld", answer);
	return 0;
}